src/actions/cart-manager.ts

```diff
@@ -1,92 +1,60 @@
-    if (productInMainCart) {
-        availableAmount = product.quantity - getAmountOfItemInMainCart(productInMainCart);
-    } else {
-        availableAmount = product.quantity;
-        productInMainCart = {id: productId, users: []};
-        mainCart.push(productInMainCart);
+export const mainCart: Map<string, Map<string, number>> = new Map();
+
+export const addToCart = async (userId: string, productId: string, desiredQuantity: number) => {
+    const productInDB = await retrieveSpecificProduct(productId);
+    if (!productInDB) {
+        throw new Error('Product not found!');
+    }
+    if (!mainCart.has(productId)) {
+        mainCart.set(productId, new Map());
+    }
+    if (mainCart.get(productId).has(userId)) {
+        throw new Error('Product already in user cart. Please remove first to add new quantity');
     }
-    if (desiredAmount <= availableAmount)
-        updateProductInMainCart(ip, desiredAmount, productInMainCart);
-    else
-        throw new Error('Quantity not enough');
-    return availableAmount - desiredAmount;
+    const availableQuantity = productInDB.quantity - getQuantityOfItemInMainCart(productId);
+    if (availableQuantity < desiredQuantity) {
+        throw new Error('Quantity not enough!');
+    }
+    mainCart.get(productId).set(userId, desiredQuantity);
+    return availableQuantity - desiredQuantity;
 };
 
 export const removeFromCart = async (userId: string, productId: string) => {
-    const product = await retrieveSpecificProduct(productId);
-    if (!product)
-        throw new Error('Item not found!');
-    const productInMainCart = findProductInMainCart(productId);
-    if (productInMainCart) {
-        removeUserFromItemInMainCart(userId, productInMainCart);
-        return product.quantity - getAmountOfItemInMainCart(productInMainCart);
-    } else
-        throw new Error('Item not found!');
-};
-
-const findProductInMainCart: (id: string) => any = (id: string) => {
-    let foundItem;
-    mainCart.forEach(cartItem => {
-        if (cartItem.id === id)
-            foundItem = cartItem;
-    });
-    return foundItem ? foundItem : null;
-};
-
-const getAmountOfItemInMainCart = (cartItem: { id: string, users: { userId: string, amount: number }[] }) => {
-    let amountInCarts = 0;
-    cartItem.users.forEach(user => amountInCarts += user.amount);
-    return amountInCarts;
-};
-
-const updateProductInMainCart = (userId: string, amount: number, cartItem: { id: string, users: { userId: string, amount: number }[] }) => {
-    let userFound = false;
-    cartItem.users.forEach(user => {
-        if (user.userId === userId) {
-            user.amount += amount;
-            userFound = true;
-        }
-    });
-    if (!userFound) {
-        const newUser = {userId, amount};
-        cartItem.users.push(newUser);
+    const productInDB = await retrieveSpecificProduct(productId);
+    if (!productInDB) {
+        throw new Error('Product not found!');
```

Comment:
whose 3 lines is duplicated (function above) - export to function

Reply:
done